### ISSUE TITLE
When is the first frame pushed to the MediaStream for captureStream(0)?

### ISSUE THREAD

**Issue by foolip** [w3c/mediacapture-fromelement#29]
Blink bug:
https://code.google.com/p/chromium/issues/detail?id=587789

Some ad-hoc tests:
http://software.hixie.ch/utilities/js/live-dom-viewer/saved/3896
http://software.hixie.ch/utilities/js/live-dom-viewer/saved/3897

The difference between the two is the order between `captureStream()` and `fillRect()`. Firefox and Chrome behave differently; regardless of the order Firefox gets the first frame and Chrome does not.

The spec says "A new frame is requested from the canvas when `frameCaptureRequested` is true and the canvas is painted" and that `frameCaptureRequested` is initially true, but doesn't define "the canvas is painted" precisely. It presumably has to do with draw commands on `CanvasRenderingContext2D` / `WebGLRenderingContext` and not painting of the canvas content on screen.

**Comment by martinthomson** [comment]
https://html.spec.whatwg.org/multipage/scripting.html#drawing-model defines what it is to "paint".  This happens not as a result of the various drawXXX function invocations, but when the actual pixels are produced.

Happy to take a PR that corrects this.

**Comment by foolip** [comment]
That looks like it's there to define how shadows, alpha, clipping and such work together, but these are rules that apply "When a shape or image is painted" and so could happen many times for each user-visible frame, even if it doesn't happen synchronously in the drawXXX calls. It looks like it could also happen zero times in some cases where the content changes, e.g. if `clearRect()` is called.

Like in https://github.com/w3c/mediacapture-fromelement/issues/27, some notion of when new frames are captured is needed, and in addition it needs to be very clear if there's an extra point of capture when `captureStream()` is called that doesn't depend on that event.

**Comment by martinthomson** [comment]
To be clear, it was assumed that this was a term of art, but clearly that isn't the case.

**Comment by foolip** [comment]
What does Firefox do here? Does it merely set `frameCaptureRequested` to true and wait for the same kind of event that https://github.com/w3c/mediacapture-fromelement/issues/27 will depend on, or does it capture the current state even if that event never happens?

**Comment by martinthomson** [comment]
@Pehrsons knows better than I, but I understand that the implementation relies solely on `frameCaptureRequested` for everything it does.

**Comment by foolip** [comment]
In https://codereview.chromium.org/1713873003/ a change is being made which sounds like it will match Gecko. For the time being, I'll assume that the spec will be updated to match this behavior.

**Comment by Pehrsons** [comment]
In Firefox we have an internal display driver that trigger paints. When this display driver fires (because of a timer or vsync) and we have a captured canvas that was drawn to from js since the last frame, and where `frameCaptureRequested` is true, we'll capture a frame and pass it to the capturing track.

We had to hook into the display driver to get access to data in a WebGL canvas with `preserveDrawingBuffer: false`.

**Comment by foolip** [comment]
A similar kind of question that has come up is whether a MediaStream holds on to its latest frame, or if it has to be connected to something (like a media element) when the frame is pushed. A test:
http://software.hixie.ch/utilities/js/live-dom-viewer/saved/3976

It definitely appears like in Gecko, a MediaStream holds on to its latest frame. Can someone confirm?

**Comment by foolip** [comment]
Also, is the hold-onto-last frame or lack of such behavior covered by any spec?

**Comment by alvestrand** [comment]
Time to rehash an old meme:
**MediaStreamTrack is a control surface**
http://alvesnotes.blogspot.se/2012/03/byte-stream-fallacy.html

It's an abstraction that allows an easy API to say what's connected to what, and possibly some transforms that need to be applied to the interconnection. It should be no more.
Given this, my opinion is that it shouldn't be required to hang on to a frame produced before the sink is known - the identity of the sink(s) is _permitted_ to influence the production process, so it should be perfectly legal for the production of the frame for the sink's consumption to happen after connection, and having a requirement that the source or places associated with it hang on to a representation that can be used with any sink until connected (potentially forever) is an onerous requirement for the benefits it provides.

That said, I also see no reason to outlaw such a hanging-on - but if I had to make a choice, I'd go with "no connection - no frame".

**Comment by foolip** [comment]
Given that we lack interoperability on this point and I discovered it with pretty simple testing with no knowledge of the implementation details, I think that this would be good to spell it out normatively so cross-browser tests can be written. I guess it would belong in https://w3c.github.io/mediacapture-main/?

(The blog post doesn't quite get at the heart of this, though. Even if "bits and bytes flowing between our objects" is the wrong model, there's still some kind of handle to a video frame moving around, and Gecko keeps an extra reference to it.)

**Comment by alvestrand** [comment]
@foolip I regard the concept of "frames" as an implementation detail.
For instance, a MediaStream that was carrying undecoded VP8 or H264 wouldn't be carrying around pointers to whole frames; it would carry around pointers to structures that linked chains of frames that ended up in an I-frame far up the stream - most of those frames are delta frames. You can't move the tail of the chain around without carrying the whole history with it.
Motion JPEG is the only codec I know about where the concept of "the frame" clearly makes sense for encoded video streams.
(This has implications for how a feature such as relaying a stream without decoding it can be implemented - one basically has to synthesize an I-frame at connect time. That's tricky.)

**Comment by foolip** [comment]
In the context of `canvas.captureStream()` the frame concept is explicit and behavior is observably different, so what the specs say has not been enough to arrive at interoperable implementations. I don't precisely know how to phrase it or even in which spec, but I think it's best to spell out somewhere the behavior that we would like to see interoperable implementations of, and I assume that's the "control surface" model.

Can anyone working on Gecko comment?

**Comment by alvestrand** [comment]
@foolip I think we agree.... when I read the canvas spec I was surprised to find that the concept of "paint" was undefined - different browsers make different choices on when to paint, and whether to paint at all for canvases that aren't being shown.
Agree fully that when producing a stream from a canvas, what happens needs to be explicitly clear. Most of my arguing is that it has to be in the canvas spec!

**Comment by foolip** [comment]
That sounds fine to me. @martinthomson, do you know how to express this in the spec? @uysalere, can you explain what Chromium's current model is, after the change in https://codereview.chromium.org/1829193003?

**Comment by uysalere** [comment]
Chromium's MediaStreamTrack implementation still does not hold onto last frame, it is barely an agent passing frames. In order to address the observable interoperability issues in canvas capture, we added functionality to request a refresh frame from Source when a new Sink is added. i.e., canvas.captureStream(0) captures a single frame and each sink attached to the stream receives that frame.

**Comment by uysalere** [comment]
To make things clear in this spec, we can add that CanvasCaptureMediaStreamTrack should provide the most recent captured frame when a new sink is added. Firefox has been doing that, and the new Chrome model does it as well. WDYT?

**Comment by Pehrsons** [comment]
Firefox remembers the last frame of a (any) video track, true (it is the track's "current" frame). That said, we [have changes coming](https://bugzilla.mozilla.org/show_bug.cgi?id=1201363) which make remembering the last frame much less obvious. This behavior might very well be going away.

**Comment by uysalere** [comment]
@pehrsons, can you clarify what do you mean by less obvious? Is it still the case that every sink is going to receive a frame? For instance, what would be the behavior on this demo: http://software.hixie.ch/utilities/js/live-dom-viewer/?saved=3976

**Comment by Pehrsons** [comment]
I had to confer with @ChiaHungTai on that. We'll be keeping our current behavior, even though this will allow us to change it more easily. It's based on the idea that every frame comes with a timestamp, and we'll remember all frames in a track where there's no next frame with a timestamp in the past.

**Comment by foolip** [comment]
> canvas.captureStream(0) captures a single frame and each sink attached to the stream receives that frame

@uysalere, I suppose this means that sink that aren't newly added will get the same frame again, which would be observable using the MediaRecorder API?

FWIW, I don't think preserving the behavior of http://software.hixie.ch/utilities/js/live-dom-viewer/?saved=3976 is super important, if it would be easy for both Firefox and Chrome to just not do anything at special when capture starts, that could work just fine I think. WDYT, @Pehrsons?

**Comment by Pehrsons** [comment]
It would be easy to implement in Firefox. The rest of your question I'll defer to @martinthomson.

**Comment by foolip** [comment]
If anyone has any hunches about what behavior is most rational or least surprising to web developers, that'd be nice too. It's not obvious to me which is better, really.

**Comment by uysalere** [comment]
@foolip, you are right. Each attached sink gets the same frame again. However, I can change it such that only the newly added sink gets it. Or, if we agree that http://software.hixie.ch/utilities/js/live-dom-viewer/?saved=3976 is ok to produce no frame, I can revert these changes. I don't have a strong opinion on this either.

**Comment by foolip** [comment]
Not producing a frame sounds OK to me. If you revert the change there isn't much long term risk because making the API more reliable shouldn't break much, while later making it less reliable is risky. So if you're unsure what's best, I'd lean towards reverting and making the spec say that there are no guarantees of this sort.

**Comment by martinthomson** [comment]
Proposed action: specify that the browser is required to remember the last painted frame for the canvas so that a new capture receives the first frame.  This is what both implementations do currently.  It conveniently avoids having nothing in a canvas - a state that could persist for a long time.

**Comment by martinthomson** [comment]
@uysalere, can you take care of this?  (I can't assign this to you, which might indicate a procedural issue.)
